[Flang] Add Sphinx man page and html support for Flang

snarang181 opened the pull request:
This patch refactors the Flang documentation CMake and Sphinx configuration to address build issues.

**CMake changes**:

- Moves the `gen_rst_file_from_td()` call out of the HTML-only block so that both `docs-flang-html` and `docs-flang-man` builds depend on the generated `FlangCommandLineReference.rst` file.

**conf.py changes**:
- Introduces `myst_parser` dependency as a required Markdown parser for both HTML and man builds.
- Introduces the correct source_suffix mapping for both .rst and .md files.
- Populates the man_pages configuration so the main index page generates a ` flang(1) `man page.

Fixes #141757

snarang181 committed:
[Flang][Docs] Add Sphinx man page support for Flang

This patch enables building Flang man pages by:

- Adding a `man_pages` entry in flang/docs/conf.py for Sphinx man builder.
- Adding a minimal `index.rst` as the master document.
- Adding placeholder `.rst` files for FIRLangRef and FlangCommandLineReference to fix toctree references.

These changes unblock builds using `-DLLVM_BUILD_MANPAGES=ON` and allow `ninja docs-flang-man` to generate `flang.1`.

Fixes #141757

snarang181 commented:
@pawosm-arm -- Sorry to tag you explicitly, I do not have access to formally request reviews using the Reviewers section.

snarang181 committed:
Remove .rst files and point conf.py to pick up .md

snarang181 committed:
While building man pages, the .md files were being used.
Due to that, the myst_parser was explictly imported.
Adding Placeholder .md files which are required by index.md

snarang181 commented:
> I think more people need to look into it. AFAIR the plan was that flang does not use the `.rst` files for documentation, it was supposed to rely on the markdown files entirely.

@pawosm-arm, thanks for looking into it. I got rid of the `.rst` files. When I added the explicit source suffix, it seemed like the man-page build needed the parser for the .md files. With the `myst_parser` library not being imported, I was hitting parsing errors of the .md files, so I explicitly changed the requirement. With that and adding placeholder missing `.md` files, I can generate a man doc at `${BUILD_DIR}/tools/flang/docs/man/flang.1`

kiranchandramohan commented:
`ninja flang-doc` might be generating files like `FIRLangRef.md`. Would adding these files manually affect that?
Note that the HTML files generated from these are displayed currently in the following pages:
https://flang.llvm.org/docs/
https://flang.llvm.org/docs/FIRLangRef.html
https://flang.llvm.org/docs/FlangCommandLineReference.html

snarang181 commented:
> https://reviews.llvm.org/D128650

@kiranchandramohan -- this runs for me successfully and does not interfere with the placeholder file. The final `FIRLangRef.md` that is produced in the build artifacts has been generated by tablegen.

kiranchandramohan commented:
> > https://reviews.llvm.org/D128650
> 
> @kiranchandramohan -- this runs for me successfully and does not interfere with the placeholder file. The final `FIRLangRef.md` that is produced in the build artifacts has been generated by tablegen.

What about the man pages and its contents. Is it the empty file or the one produced by tablegen?

kiranchandramohan commented:
> > > > https://reviews.llvm.org/D128650
> > > 
> > > 
> > > @kiranchandramohan -- this runs for me successfully and does not interfere with the placeholder file. The final `FIRLangRef.md` that is produced in the build artifacts has been generated by tablegen.
> > 
> > 
> > What about the man pages and its contents. Is it the empty file or the one produced by tablegen?
> 
> As I mentioned the man page produced at `${BUILD_DIR}/tools/flang/docs/man/flang.1` looks like the one produced by tablegen. The `FIRLangRef.md` looks to be completely overwritten in the build directory. For the other placeholder file, `FlangCommandLineReference.md`, there is no fallout either.

Can we do something similar to https://reviews.llvm.org/D128650 to avoid needing dummy `FIRLangRef.md` and `FlangCommandLineReference`?

snarang181 committed:
Remove placeholder .md files

tarunprabhu commented:
> @kiranchandramohan, I removed the placeholder `.md` files and the man pages get generated now. Just a warning of the form `WARNING: toctree contains reference to nonexisting document 'FIRLangRef'` `WARNING: toctree contains reference to nonexisting document 'FlangCommandLineReference'` This is probably OK?

It might not be ok. If the buildkite fails, as it does [here](https://github.com/llvm/llvm-project/actions/runs/15324450618/job/43119288153?pr=141882), it will probably break buildbots too. I am not sure what is causing this though.

snarang181 committed:
Enable docs-flang-html to build

snarang181 committed:
Modify CMake to build man without warnings

snarang181 committed:
Minor fix for building html version
copy-flang-src-docs builds Source/ directory
and in fixing, did not include it in the HTML target deps

snarang181 commented:
@kiranchandramohan, @tarunprabhu, @pawosm-arm
Hi guys, I think this PR is ready for review. We can successfully build both HTML and man targets with these changes. Let me know how it looks.
Thanks.

pawosm-arm commented:
Not sure what is the ultimate scope of this PR, the flang manpage is indeed being generated. There are some observations though, but I'm not sure if this is the PR which should address them. Namely:

- At the very beginning we can see this:

```
FLANG |VERSION| (IN-PROGRESS) RELEASE NOTES
          warning
```

- There are some mentions of the problems with building on Ubuntu 18. This is slightly old Linux distribution these days, I doubt many cares about those issues, and it doesn't feel like a manpage is a good place for running into details of them

- The man page is way too big, way too extensive and way too detailed. It feels like entire Flang documentation went into it. Was this intentional?

I like the fact that this page finally generates easily with no need for extensive CMake configuring.

snarang181 commented:
> * The man page is way too big, way too extensive and way too detailed. It feels like entire Flang documentation went into it. Was this intentional?

I think the man page is intended to be that way since all the documentation is built into `index.md`. If we wanted more separate man pages, we can more config entries but that did not seem to be the intended use case.

snarang181 commented:
> > > * The man page is way too big, way too extensive and way too detailed. It feels like entire Flang documentation went into it. Was this intentional?
> > 
> > 
> > I think the man page is intended to be that way since all the documentation is built into `index.md`. If we wanted more separate man pages, we can more config entries but that did not seem to be the intended use case.
> 
> What I meant is this seems unusual. Compare these:
> 
> ```
> $ ls -la *lang*
> -rw-r--r-- 1 user01 users    28437 May 31 10:26 clang.1
> -rw-r--r-- 1 user01 users     1450 May 31 10:20 clang-tblgen.1
> -rw-r--r-- 1 user01 users  1199536 May 31 10:26 flang.1
> ```

Ah ok, I see your point. I will defer this to someone who knows the expected size or if it really is meant to be that big.

kiranchandramohan commented on flang/docs/CMakeLists.txt:
```diff
@@ -105,32 +103,35 @@ endfunction()
-      add_sphinx_target(man flang)
+      add_sphinx_target(man flang SOURCE_DIR "${CMAKE_CURRENT_BINARY_DIR}/Source")
+      add_dependencies(docs-flang-man gen-FlangCommandLineReference.rst)
+      add_dependencies(docs-flang-man copy-flang-src-docs)
```
`copy-flang-src-docs` seems to depend on `gen-FlangCommandLineReference.rst` in the definition above. So is it required to add the dependence on `gen-FlangCommandLineReference.rst` again?

Also, where is `gen-FlangCommandLineReference.rst` defined?

kiranchandramohan reviewed (approved):
Thanks for the changes. This looks OK to me. I have a question inline.

LGTM. Please wait for @tarunprabhu and @pawosm-arm

snarang181 committed:
Remove redundant dependence in CMake config

snarang181 commented:
> > > I'm in two minds here. I've tried this PR after forced push, yet nothing really has changed from my previous comment (there's still this `warning` word visible, and.
> 
> > 
> 
> > By the warning, you mean inside the contents of the man page right? Not the build process giving warning to nonexistent documents.
> 
> 
> 
> Yes, as I quoted it above:
> 
> 
> 
> ```
> 
> FLANG(1)                               Flang                              FLANG(1)
> 
> 
> 
> NAME
> 
>        flang - Flang Documentation
> 
> 
> 
>        Flang  is  LLVM's  Fortran frontend that can be found here. It is often reâ
> 
>        ferred to as "LLVM Flang" to differentiate itself from  "Classic  Flang"  -
> 
>        these are two separate and independent Fortran compilers. LLVM Flang is unâ
> 
>        der active development. While it is capable of generating executables for a
> 
>        number  of  examples,  some functionality is still missing. See Getting Inâ
> 
>        volved for tips on how to get in touch with us and to learn more about  the
> 
>        current status.
> 
> 
> 
> FLANG |VERSION| (IN-PROGRESS) RELEASE NOTES
> 
>           warning
> 
> 
> 
>           These  are  in-progress  notes  for the upcoming LLVM |version| release.
> 
>           Release notes for previous releases can be found on the Download Page.
> 
> ```

Ok, I see your point. I think a follow up PR to iron out the contents of the man page should be the course of action since the issue this fix links to is for the build process only. The changes might get too convoluted; what do you think?

pawosm-arm commented:
> Something to consider: maybe we should introduce the use of something like this? https://github.com/cpuguy83/go-md2man

This seems somewhat simpler: https://kristaps.bsd.lv/lowdown

pawosm-arm commented:
> > Something to consider: maybe we should introduce the use of something like this? https://github.com/cpuguy83/go-md2man
> 
> This seems somewhat simpler: https://kristaps.bsd.lv/lowdown

From the other side, the manpage flang-doc generates from markdown demonstrates that we are able to make markdown to manpage conversion without engaging any external tools. What we lack is an input which could establish the correct structure, something that clang has (see https://github.com/llvm/llvm-project/blob/main/clang/docs/CommandGuide/clang.rst) but in the markdown format. I'd rather see this PR introducing a stub for such content rather than generating a substantial in size content made of something which isn't a manpage.

snarang181 commented:
> > > Something to consider: maybe we should introduce the use of something like this? https://github.com/cpuguy83/go-md2man
> 
> > 
> 
> > This seems somewhat simpler: https://kristaps.bsd.lv/lowdown
> 
> 
> 
> From the other side, the manpage flang-doc generates from markdown demonstrates that we are able to make markdown to manpage conversion without engaging any external tools. What we lack is an input which could establish the correct structure, something that clang has (see https://github.com/llvm/llvm-project/blob/main/clang/docs/CommandGuide/clang.rst) but in the markdown format. I'd rather see this PR introducing a stub for such content rather than generating a substantial in size content made of something which isn't a manpage.

Sounds good. 
Can we agree on a stub for the purposes of this PR? 
Cc @kiranchandramohan @tarunprabhu

snarang181 committed:
Add Placeholder man page template for flang docs

snarang181 committed:
Modify CMake to support man page build
Separate html and man page builds in CMakeLists.txt

pawosm-arm commented on flang/docs/conf.py:
```diff
@@ -10,6 +10,7 @@
 # serve to show the default.
 
 from datetime import date
+
```
accidental empty line

pawosm-arm commented on flang/docs/index.rst:
```diff
@@ -0,0 +1,21 @@
+
+https://flang.llvm.org/docs/
+
+TODO: This man page currently shows FlangCommandLineReference this with finalized man page content.
```
I suppose the `this` word is not needed here, makes it harder to read.

snarang181 committed:
Address suggestions

pawosm-arm commented:
I'd accept this, but the others must agree with the introduction of the rst files next to md files.

snarang181 commented:
> I'd accept this, but the others must agree with the introduction of the rst files next to md files.

@tarunprabhu, @kiranchandramohan, any comments or suggestions?

tarunprabhu commented:
> I'd accept this, but the others must agree with the introduction of the rst files next to md files.

I will defer to @kiranchandramohan on the introduction of `.rst` files.

snarang181 commented:
> We may want to consider auto-generating a command-line reference and having that in the man page, but that is a broader, and potentially longer, discussion that should not hold up this PR.

@tarunprabhu, IIUC, the `FlangCommandLineReference.rst` is generated from tablegen in the build process and we link that file in the generated man page for now.

tarunprabhu commented on flang/docs/index.rst:
```diff
@@ -0,0 +1,21 @@
+
+https://flang.llvm.org/docs/
+
+TODO: This man page currently shows FlangCommandLineReference, will be ultimately replaced with finalized man page content.
```
I don't think this should be visible in the actual man page. It feels like a comment for developers, so may be something like this instead?

```suggestion
..
  The placeholder text "FlangCommandLineReference" below should eventually be replaced with the actual man page contents.
```

tarunprabhu commented on flang/docs/CMakeLists.txt:
```diff
@@ -105,32 +101,63 @@ endfunction()
+                "${CMAKE_CURRENT_BINARY_DIR}/Source/index.rst"
+        COMMAND "${CMAKE_COMMAND}" -E copy
+                "${CMAKE_CURRENT_BINARY_DIR}/Source/FlangCommandLineReference.rst"
+                "${CMAKE_CURRENT_BINARY_DIR}/Source/FlangCommandLineReference.rst"
```
Is this correct? Unless I am missing something, this seems to be copying the file onto itself

snarang181 committed:
Remove redundant command in CMake
Modify man page doc

kiranchandramohan commented:
> > I'd accept this, but the others must agree with the introduction of the rst files next to md files.
> 
> I will defer to @kiranchandramohan on the introduction of `.rst` files.

I am OK if it can be put in a sub-directory like clang (docs/CommandGuide).

snarang181 committed:
Move index.rst to docs/CommandGuide

snarang181 committed:
Adjust path to CommandLineReference file

snarang181 committed:
relative path change in removing of CommandLineGuide

kiranchandramohan reviewed (approved):
LGTM. Thanks for all the changes. Please wait for @tarunprabhu or @pawosm-arm

pawosm-arm commented:
I'm trying to build it, but it fails, I presume due to warnings being treated as errors as I can't see any actual errors in the log. Does it build for you?:

```
WARNING: multiple files found for the document "index": index.md, index.rst
```

snarang181 commented:
> I'm trying to build it, but it fails, I presume due to warnings being treated as errors as I can't see any actual errors in the log. Does it build for you?:
> 
> 
> 
> ```
> 
> WARNING: multiple files found for the document "index": index.md, index.rst
> 
> ```

Are you building man or HTML?

pawosm-arm commented:
> > > > I'm trying to build it, but it fails, I presume due to warnings being treated as errors as I can't see any actual errors in the log. Does it build for you?:
> > > > ```
> > > > 
> > > > WARNING: multiple files found for the document "index": index.md, index.rst
> > > > ```
> > > 
> > > 
> > > Are you building man or HTML?
> > 
> > 
> > This is a simple change I've added to the build script (I presume this is how it will be used by the others too):
> > ```
> >     -DLLVM_BUILD_DOCS=ON
> >     -DLLVM_ENABLE_SPHINX=ON
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This results in the following lines being generated in the `CMakeCache.txt` file:
> > ```
> > //Generate build targets for the Flang docs.
> > FLANG_INCLUDE_DOCS:BOOL=ON
> > 
> > //HTML documentation install directory for flang
> > FLANG_INSTALL_SPHINX_HTML_DIR:STRING=share/doc/LLVM/flang/html
> > 
> > //Output standalone HTML files
> > SPHINX_OUTPUT_HTML:BOOL=ON
> > 
> > //Output man pages
> > SPHINX_OUTPUT_MAN:BOOL=ON
> > 
> > //When building documentation treat warnings as errors
> > SPHINX_WARNINGS_AS_ERRORS:BOOL=ON
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > From this you can read I'm building both.
> 
> So I have both enabled too according to my local `CMakeCache.txt` and I am able to build both targets. This error was seen before, just as a sanity check, do you have a clean build?

I did a clean built after cherry-picking your recent commits to this PR over the same base commit (to which I'm doing git-reset each time you do a force-push). I don't know if they were seen before, as the build was passing before those most recent cherry-picks so not much has changed really.

samanara-qti committed:
Separate out build directories for man and html builds

samanara-qti committed:
Modify build script to use HTML routes

snarang181 commented:
> > > > > I'm trying to build it, but it fails, I presume due to warnings being treated as errors as I can't see any actual errors in the log. Does it build for you?:
> > > > > ```
> > > > > 
> > > > > WARNING: multiple files found for the document "index": index.md, index.rst
> > > > > ```
> > > > 
> > > > 
> > > > Are you building man or HTML?
> > > 
> > > 
> > > This is a simple change I've added to the build script (I presume this is how it will be used by the others too):
> > > ```
> > >     -DLLVM_BUILD_DOCS=ON
> > >     -DLLVM_ENABLE_SPHINX=ON
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > This results in the following lines being generated in the `CMakeCache.txt` file:
> > > ```
> > > //Generate build targets for the Flang docs.
> > > FLANG_INCLUDE_DOCS:BOOL=ON
> > > 
> > > //HTML documentation install directory for flang
> > > FLANG_INSTALL_SPHINX_HTML_DIR:STRING=share/doc/LLVM/flang/html
> > > 
> > > //Output standalone HTML files
> > > SPHINX_OUTPUT_HTML:BOOL=ON
> > > 
> > > //Output man pages
> > > SPHINX_OUTPUT_MAN:BOOL=ON
> > > 
> > > //When building documentation treat warnings as errors
> > > SPHINX_WARNINGS_AS_ERRORS:BOOL=ON
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > From this you can read I'm building both.
> > 
> > 
> > So I have both enabled too according to my local `CMakeCache.txt` and I am able to build both targets. This error was seen before, just as a sanity check, do you have a clean build?
> 
> I did a clean build after cherry-picking your recent commits to this PR over the same base commit (to which I'm doing git-reset each time you do a force-push). I don't know if these messages were seen before, as the build was passing before those most recent cherry-picks so not much has changed really.

No, you are right, I hadn't done the builds back to back, just independently. I have a fix.

samanara-qti committed:
Explicitly copy generated rst file for HTML build

snarang181 commented:
@pawosm-arm, can you test it out?

snarang181 commented:
> > > I'm trying to build it, but it fails, I presume due to warnings being treated as errors as I can't see any actual errors in the log. Does it build for you?:
> > > ```
> > > 
> > > WARNING: multiple files found for the document "index": index.md, index.rst
> > > ```
> > 
> > 
> > Are you building man or HTML?
> 
> This is a simple change I've added to the build script (I presume this is how it will be used by the others too):
> 
> ```
>     -DLLVM_BUILD_DOCS=ON
>     -DLLVM_ENABLE_SPHINX=ON
> ```
> 
> This results in the following lines being generated in the `CMakeCache.txt` file:
> 
> ```
> //Generate build targets for the Flang docs.
> FLANG_INCLUDE_DOCS:BOOL=ON
> 
> //HTML documentation install directory for flang
> FLANG_INSTALL_SPHINX_HTML_DIR:STRING=share/doc/LLVM/flang/html
> 
> //Output standalone HTML files
> SPHINX_OUTPUT_HTML:BOOL=ON
> 
> //Output man pages
> SPHINX_OUTPUT_MAN:BOOL=ON
> 
> //When building documentation treat warnings as errors
> SPHINX_WARNINGS_AS_ERRORS:BOOL=ON
> ```
> 
> From this you can read I'm building both.
@pawosm-arm
 **What was going on:** When we did a `ninja docs-flang-man` followed by `ninja docs-flang-html`, there were two refs to an `index` file -- one with a `.md` and another with an `.rst`. This was mainly because we had the same `CMAKE_BINARY_DIR/Source` dir. So I have separated the `Source` directories for both the builds to have independent workflows (this can be removed of course once we unify the index file).

pawosm-arm commented:
> @pawosm-arm, can you test it out?

Yes I'll trigger another build today.

pawosm-arm commented:
> > @pawosm-arm, can you test it out?
> 
> Yes I'll trigger another build today.

It looks much better now. Although the `Introduction` section looks weird (the way it starts going directly to listing the command line options), but I tend to agree that it's a good for the starters and the proper refinement should happened later and should wait for the contributors to pick it up. For now it looks as such:

```
FLANG(1)                                Flang                               FLANG(1)

NAME
       flang - Flang Documentation (In Progress)

       NOTE:
          This man page is under development.

       For full documentation, please see the online HTML docs:

       https://flang.llvm.org/docs/

                                           ----

FLANG COMMAND LINE REFERENCE
   Flang command line argument reference
       o Introduction

       o Actions

       o Compilation options

         o Preprocessor options

           o Include path management

           o Dumping preprocessor state

         o Diagnostic options

         o Target-independent compilation options

           o Common Offloading options

           o HIP options

         o Target-dependent compilation options

           o AARCH64

           o X86

           o LoongArch

         o Optimization level

         o Debug information generation

           o Kind and level of debug information

             o Debug level

       o Linker options

   Introduction
       -B<prefix>

       Search $prefix$file for executables, libraries, and data files. If $prefix is
       a directory, search $prefix/$file

       -Qunused-arguments

       Don't emit warning for unused driver arguments

       --config=<file>, --config <arg>
 Manual page flang(1) line 1 (press h for help or q to quit)
```

snarang181 commented:
Whoever has writeaccess, please merge this.

pawosm-arm commented:
> Whoever has writeaccess, please merge this.

Done
